### PR TITLE
Allow for `identifier` dicts in User Interactive Auth dicts

### PR DESCRIPTION
**Pull request by anoadragon453**
A stab at fixing https://github.com/matrix-org/synapse/issues/5665

Allows `identifier` dicts to be used in UIA flows and relaxes the requirement of a `user` parameter, which has since been deprecated. Relevant spec: https://matrix.org/docs/spec/client_server/r0.6.1#identifier-types

**Warning:** This deals with UIA and login and may break both - careful review required.

Requires: https://github.com/matrix-org/synapse/pull/7687

Should be reviewable commit-by-commit.

Note: This PR reworks the end of `_check_auth_dict`, which was essentially the place where `m.login.password` UI Auth steps were handled. In a future PR, it may be cleaner to move that code into an instance of `UserInteractiveAuthChecker` in [synapse/handlers/ui_auth/checkers.py](https://github.com/matrix-org/synapse/blob/a3fbc23c39c0366392fd51faf0b1696f1f1d21c7/synapse/handlers/ui_auth/checkers.py#L29).

### CODE REVIEW

File: synapse/handlers/auth.py [hunk 1]
```diff
@@ -520,12 +520,18 @@ async def _check_auth_dict(
...
             return res
 
-        # build a v1-login-style dict out of the authdict and fall back to the
-        # v1 code
-        user_id = authdict.get("user")
+        # Retrieve the user ID from the authdict
+
+        # Check for an identifier dict
+        identifier = authdict.get("identifier", {})
+        if "type" in identifier and identifier["type"] == "m.id.user":
```
Comment (clokep) [inline synapse/handlers/auth.py:527]: You should be able to just get it all at once here without checking.

```suggestion
        if identifier.get("type") == "m.id.user":
```

Comment (anoadragon453) [reply synapse/handlers/auth.py:527]: Nice, thanks!

File: synapse/handlers/auth.py [hunk 2]
```diff
@@ -45,12 +45,93 @@
...
+    # Ensure the identifier has a type
+    if "type" not in submission["identifier"]:
+        raise SynapseError(
+            400, "'identifier' dict has no key 'type'", errcode=Codes.MISSING_PARAM,
+        )
+
+
+def login_id_phone_to_thirdparty(identifier: Dict[str, str]):
+    """Convert a phone login identifier type to a generic threepid identifier. Modifies
+    the identifier dict in place
```
Comment (clokep) [inline synapse/handlers/auth.py:100]: Why the change to modify this in place? The original version of this returned a new dict.

(Also a minor thing, but the commit description just says that code is moved, not modified.)

File: synapse/handlers/auth.py [hunk 3]
```diff
@@ -45,12 +45,93 @@
...
+    https://matrix.org/docs/spec/client_server/r0.6.1#identifier-types
+    Args:
+        submission: The client dict to convert. Passed by reference and modified
+    Raises:
+        SynapseError: if the dict contains a "medium" parameter that is anything other than
+            "email"
+    """
+    if "user" in submission:
+        submission["identifier"] = {"type": "m.id.user", "user": submission["user"]}
+        del submission["user"]
```
Comment (clokep) [inline synapse/handlers/auth.py:71]:
```suggestion
        submission["identifier"] = {"type": "m.id.user", "user": submission.pop("user")}
```

Comment (clokep) [reply synapse/handlers/auth.py:71]: (A similar change can be made below.)

File: synapse/handlers/auth.py [hunk 4]
```diff
@@ -45,12 +45,93 @@
...
+        del submission["address"]
+
+    # We've converted valid, legacy login submissions to an identifier. If the
+    # dict still doesn't have an identifier, it's invalid
+    if "identifier" not in submission:
+        raise SynapseError(
+            400,
+            "Missing 'identifier' parameter in login submission",
+            errcode=Codes.MISSING_PARAM,
+        )
```
Comment (clokep) [inline synapse/handlers/auth.py:89]: It probably doesn't make sense to use here, but this is pretty much `assert_params_in_dict`.

File: synapse/handlers/auth.py [hunk 5]
```diff
@@ -45,12 +45,93 @@
...
 
+def client_dict_convert_legacy_fields_to_identifier(
+    submission: Dict[str, Union[str, Dict]]
+):
+    """Take a legacy-formatted login submission or User-Interactive Authentication dict and
+    updates it to feature an identifier dict instead.
+    Providing user-identifying information at the top-level of a login or UIA submission is
+    now deprecated and replaced with identifiers:
+    https://matrix.org/docs/spec/client_server/r0.6.1#identifier-types
+    Args:
```
Comment (clokep) [inline synapse/handlers/auth.py:63]:
```suggestion
    """
    Convert a legacy-formatted login submission to an identifier dict.

    Legacy login submissions (used in both login and user-interactive authentication)
    provide user-identifying information at the top-level instead of in an `indentifier`
    property. This is now deprecated and replaced with identifiers:
    https://matrix.org/docs/spec/client_server/r0.6.1#identifier-types

    Args:
```

File: synapse/handlers/auth.py [hunk 6]
```diff
@@ -45,12 +45,93 @@
...
+        submission: The client dict to convert. Passed by reference and modified
+    Raises:
+        SynapseError: if the dict contains a "medium" parameter that is anything other than
+            "email"
```
Comment (clokep) [inline synapse/handlers/auth.py:67]: This comment seems wrong? It raises if the format is invalid.

File: synapse/handlers/auth.py [hunk 7]
```diff
@@ -603,6 +603,87 @@ async def _check_auth_dict(
         (canonical_id, callback) = await self.validate_login(user_id, authdict)
         return canonical_id
 
+    async def username_from_identifier(
+        self, identifier: Dict[str, str], password: Optional[str] = None
+    ) -> Optional[str]:
+        """Given a dictionary containing an identifier from a client, extract the
+        possibly unqualified username of the user that it identifies. Does *not*
+        guarantee that the user exists.
+        If this identifier dict contains a threepid, we attempt to ask password
```
Comment (clokep) [inline synapse/handlers/auth.py:640]: Please add blank lines between the different sections in docstrings, it makes it much easier to read!

File: synapse/rest/client/v1/login.py [hunk 8]
```diff
@@ -144,99 +141,49 @@ async def _do_other_login(self, login_submission):
...
-                canonical_user_id,
-                callback_3pid,
-            ) = await self.auth_handler.check_password_provider_3pid(
-                medium, address, login_submission["password"]
+        # Check whether this attempt uses a threepid, if so, check if our failed attempt
+        # ratelimiter allows another attempt at this time
+        medium, address = (
+            login_submission.get("medium"),
+            login_submission.get("address"),
+        )
```
Comment (clokep) [inline synapse/rest/client/v1/login.py:150]: I don't think it is clearer to have this be a tuple that gets unpacked.

File: synapse/rest/client/v1/login.py [hunk 9]
```diff
@@ -144,99 +141,49 @@ async def _do_other_login(self, login_submission):
...
+        if medium and address:
+            self._failed_attempts_ratelimiter.ratelimit(
+                (medium.lower(), address.lower()), update=False
```
Comment (clokep) [inline synapse/rest/client/v1/login.py:153]: The lower-casing here seems to be a change in behavior? Or maybe I can't follow the logic...

File: synapse/rest/client/v1/login.py [hunk 10]
```diff
@@ -144,99 +141,49 @@ async def _do_other_login(self, login_submission):
...
-                raise SynapseError(400, "Invalid thirdparty identifier")
-
-            if medium == "email":
-                # For emails, transform the address to lowercase.
-                # We store all email addreses as lowercase in the DB.
-                # (See add_threepid in synapse/handlers/auth.py)
-                address = address.lower()
-
-            # We also apply account rate limiting using the 3PID as a key, as
-            # otherwise using 3PID bypasses the ratelimiting based on user ID.
```
Comment (clokep) [inline synapse/rest/client/v1/login.py:205]: This comment was helpful, I think.

File: synapse/rest/client/v1/login.py [hunk 11]
```diff
@@ -144,99 +141,49 @@ async def _do_other_login(self, login_submission):
...
+                    (medium.lower(), address.lower())
                 )
-                # We mark that we've failed to log in here, as
-                # `check_password_provider_3pid` might have returned `None` due
-                # to an incorrect password, rather than the account not
-                # existing.
-                #
-                # If it returned None but the 3PID was bound then we won't hit
-                # this code path, which is fine as then the per-user ratelimit
-                # will kick in below.
```
Comment (clokep) [inline synapse/rest/client/v1/login.py:239]: This comment seems more helpful than the new version.

File: synapse/rest/client/v1/login.py [hunk 12]
```diff
@@ -144,99 +141,49 @@ async def _do_other_login(self, login_submission):
...
+            self._failed_attempts_ratelimiter.ratelimit(
+                (medium.lower(), address.lower()), update=False
             )
-            if canonical_user_id:
-                # Authentication through password provider and 3pid succeeded
 
-                result = await self._complete_login(
-                    canonical_user_id, login_submission, callback_3pid
-                )
-                return result
```
Comment (clokep) [inline synapse/rest/client/v1/login.py:221]: This code path seems to go through an additional `self.auth_handler.validate_login`, I don't know if that's OK or not.

File: synapse/rest/client/v1/login.py [hunk 13]
```diff
@@ -144,99 +141,49 @@ async def _do_other_login(self, login_submission):
...
-        # else, we haven't understood it.
-        if identifier["type"] != "m.id.user":
-            raise SynapseError(400, "Unknown login identifier type")
-        if "user" not in identifier:
-            raise SynapseError(400, "User identifier is missing 'user' key")
-
-        if identifier["user"].startswith("@"):
-            qualified_user_id = identifier["user"]
-        else:
-            qualified_user_id = UserID(identifier["user"], self.hs.hostname).to_string()
```
Comment (clokep) [inline synapse/rest/client/v1/login.py:255]: This logic seems to have disappeared.

File: synapse/handlers/auth.py [hunk 14]
```diff
@@ -888,10 +889,10 @@ async def validate_login(
             LoginError if there was an authentication problem.
         """
 
-        if username.startswith("@"):
-            qualified_user_id = username
-        else:
-            qualified_user_id = UserID(username, self.hs.hostname).to_string()
+        # We need a fully qualified User ID for some method calls here
+        qualified_user_id = username
```
Comment (clokep) [inline synapse/handlers/auth.py:893]: Can we just rename the input? (Or rename all users of `qualified_user_id`?)

File: tests/rest/client/v2_alpha/test_auth.py [hunk 15]
```diff
@@ -38,11 +38,6 @@ def check_auth(self, authdict, clientip):
         return succeed(True)
 
 
-class DummyPasswordChecker(UserInteractiveAuthChecker):
-    def check_auth(self, authdict, clientip):
-        return succeed(authdict["identifier"]["user"])
```
Comment (clokep) [inline tests/rest/client/v2_alpha/test_auth.py:43]: Is this just not used?

File: tests/rest/client/v1/test_login.py [hunk 16]
```diff
@@ -273,9 +273,7 @@ def _delete_device(self, access_token, user_id, password, device_id):
 
         auth = {
             "type": "m.login.password",
-            # https://github.com/matrix-org/synapse/issues/5665
-            # "identifier": {"type": "m.id.user", "user": user_id},
-            "user": user_id,
+            "identifier": {"type": "m.id.user", "user": user_id},
```
Comment (clokep) [inline tests/rest/client/v1/test_login.py:276]: We likely still want a test for the legacy login type.

File: synapse/handlers/auth.py (hunk 2 above)
Comment (anoadragon453) [reply synapse/handlers/auth.py:100]: I'm actually not sure why the function signature change was necessary...

Have modified back to return a new dict.

File: synapse/handlers/auth.py (hunk 4 above)
Comment (anoadragon453) [reply synapse/handlers/auth.py:89]: True yeah. Actually now that this is used for UI Auth I don't think it makes sense to still say "login submission".

And if I remove that, we just end up with `assert_params_in_dict` :)

File: synapse/handlers/auth.py (hunk 6 above)
Comment (anoadragon453) [reply synapse/handlers/auth.py:71]: Ah, that was a check I added, as that's technically [required by the spec for login](https://matrix.org/docs/spec/client_server/unstable#post-matrix-client-r0-login) although it's a bit of a weird requirement if `identifier` is allowed to be `m.id.phone`.

Will remove the comment though.

File: synapse/handlers/auth.py (hunk 7 above)
Comment (anoadragon453) [reply synapse/handlers/auth.py:637]: I agree with you, not sure why I did that in the first place :thinking:

File: synapse/rest/client/v1/login.py (hunk 8 above)
Comment (anoadragon453) [reply synapse/rest/client/v1/login.py:150]: Versus:

```
medium = login_submission.get("medium")
address = login_submission.get("address")
```

?

File: synapse/rest/client/v1/login.py (hunk 9 above)
Comment (anoadragon453) [reply synapse/rest/client/v1/login.py:153]: Ah! Yes it is, and apologies for not calling it out during the commit.

This would've allowed someone to bypass the ratelimiter by specifying `eMail` or some other combination of capitalisation.

However, now that I'm testing it, this trickery will get caught later on in the flow anyways, and return an immediate 403 before trying anything CPU-heavy like password hashing.

So it's not too sensitive - but I am going to pull it out of here into another PR as I think it's worthwhile, but needs more careful thought.

File: synapse/rest/client/v1/login.py (hunk 10 above)
Comment (anoadragon453) [reply synapse/rest/client/v1/login.py:205]: Hmm, I think calling out the 3PID is being used as a key here as opposed to user_id elsewhere is useful, but the second half is sort of implied in the new code I think?

File: synapse/rest/client/v1/login.py (hunk 11 above)
Comment (anoadragon453) [reply synapse/rest/client/v1/login.py:239]: Hmm, I think things have slightly changed now though (which may or may not be an issue as well).

> # If it returned None but the 3PID was bound then we won't hit
> # this code path,

This is no longer true. If `None` is returned by `username_from_identifier` and a `medium` and `address` was used, then we will check the ratelimiter. This *sounds* sensible, but it is a change from before, where it seems like the code would rate limit on the User ID instead.

File: synapse/rest/client/v1/login.py (hunk 13 above)
Comment (anoadragon453) [reply synapse/rest/client/v1/login.py:255]: It moved to `AuthHandler.validate_login`.

**Comment by anoadragon453** [comment]
I've noticed that the User ID ratelimiter now ratelimits on `username`, rather than `qualified_user_id`, which means that technically one can beat the ratelimiter by switching from localpart to full user_id. Will need to figure out a way to restore the original behaviour.

File: synapse/rest/client/v1/login.py (hunk 12 above)
Comment (anoadragon453) [reply synapse/rest/client/v1/login.py:221]: Aha, yes good spot.

I believe this originally did not as we have already validated the password via the password auth provider, so there's no need to `validate_login` at that point (which will itself ask password auth providers with a user_id/password combo).

Considering this would change Synapse's behaviour related to third-party code we should probably try to get the behaviour to match.

File: synapse/handlers/auth.py (hunk 14 above)
Comment (anoadragon453) [reply synapse/handlers/auth.py:890]: `username` could be a localpart here, however we need both the original input, as well as qualified user IDs for this code, so we create two variables.

File: tests/rest/client/v2_alpha/test_auth.py (hunk 15 above)
Comment (anoadragon453) [reply tests/rest/client/v2_alpha/test_auth.py:43]: It was used by https://github.com/matrix-org/synapse/pull/7438/commits/18071156e49d49847b87a418db09d04346ee9b16#diff-5fc780094ae3ec668024efd505af6a3dL170

But this checker was a patch to allow for identifiers being passed to `AuthHandler.check_auth`, but this PR removes the need for it as it providers that functionality.

File: tests/rest/client/v1/test_login.py (hunk 16 above)
Comment (anoadragon453) [reply tests/rest/client/v1/test_login.py:276]: Hm, so the `login` unittest function still uses the old method: https://github.com/matrix-org/synapse/blob/c2e1a2110fbe9ead26b4ecbb1afd504ed035a04d/tests/unittest.py#L512-L518 So it's definitely still tested.

But it may be a better idea to update this to the new format and just have a specific test for the old behaviour.

File: synapse/handlers/auth.py (hunk 14 above)
Comment (anoadragon453) [reply synapse/handlers/auth.py:890]: Descriptions of all the password auth provider functions are here: https://github.com/matrix-org/synapse/blob/develop/docs/password_auth_providers.md

`check_auth` is supposed to be what the user provided, localpart or not.

File: synapse/rest/client/v1/login.py (hunk 8 above)
Comment (clokep) [reply synapse/rest/client/v1/login.py:150]: Yep!

File: synapse/rest/client/v1/login.py (hunk 9 above)
Comment (clokep) [reply synapse/rest/client/v1/login.py:153]: Yes, please do! 👍

File: synapse/handlers/auth.py (hunk 14 above)
Comment (clokep) [reply synapse/handlers/auth.py:890]: I was not clear in my comment, sorry.

I meant that I found it odd for a function to have a `username` parameter which is used in one place (to save it into a separate variable). That's why I was asking about renaming all uses of `qualified_user_id` to `username` since we only seem to use it once.

Anyway, this change is really just a style change and renaming all of that would make the diff bigger, so ignore this comment and let's leave it.

File: synapse/rest/client/v1/login.py (hunk 10 above)
Comment (clokep) [reply synapse/rest/client/v1/login.py:205]: I think this code is confusing enough that it is not bad to be explicit.

File: synapse/rest/client/v1/login.py (hunk 11 above)
Comment (clokep) [reply synapse/rest/client/v1/login.py:239]: Hmm...I think I was thinking the part about the incorrect password vs. the account not existing seems useful to keep (if that's still true).

File: synapse/rest/client/v1/login.py (hunk 13 above)
Comment (clokep) [reply synapse/rest/client/v1/login.py:255]: Ah, actually this was always in `validate_login`, I think? So this was just duplicated?

File: synapse/rest/client/v1/login.py (hunk 8 above)
Comment (anoadragon453) [reply synapse/rest/client/v1/login.py:150]: Yeah, I agree :+1:

**Comment by anoadragon453** [comment]
Closing this for now as the code is just bitrotting. Though I do plan to break this up further and apply it against `$current_mainline` once I find some time to do so.
